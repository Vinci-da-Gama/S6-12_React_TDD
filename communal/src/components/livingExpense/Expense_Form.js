import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { Button, Alert } from 'reactstrap';

class ExpenseForm extends Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange(e) {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onAmountChange(e) {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    }

    onDateChange(createdAt) {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    }

    onFocusChange({ focused }) {
        this.setState(() => ({ calendarFocused: focused }));
    }

    onNoteChange(e) {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onFormSubmit(e) {
        e.preventDefault();
        const error = 'Please provide description and amount.';
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }
    
    render () {
        return (
            <div>
                { this.state.error && <Alert color="danger">{this.state.error}</Alert> }
                <form noValidate name="expenseForm" onSubmit={ (event) => { this.onFormSubmit(event) } }>
                    <input type="text" name="ef_description" placeholder="Description..." autoFocus
                        value={this.state.description} onChange={ (event) => { this.onDescriptionChange(event) }} />
                    <input type="text" name="ef_Amount" placeholder="Amount..."
                        value={this.state.amount} onChange={ (event) => { this.onAmountChange(event) } } />
                    <SingleDatePicker date={this.state.createdAt} focused={this.state.calendarFocused}
                        onDateChange={ this.onDateChange } onFocusChange={ this.onFocusChange }
                        numberOfMonths={1} isOutsideRange={ () => false } />
                    <textarea name="ef_note" rows="5"
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={ (event) => this.onNoteChange(event) }></textarea>
                    <Button color="success" outline block>
                        Add_Expense
                    </Button>
                </form>
            </div>
        );
    }
};

export default ExpenseForm;