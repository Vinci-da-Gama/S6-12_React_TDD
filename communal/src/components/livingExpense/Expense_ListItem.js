import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { ListGroup, ListGroupItem, Badge, Label } from 'reactstrap';

const ExpenseListItem = ({ id, description, amount, createdAt }) => {
    return (
        <div>
            <ListGroup>
                <ListGroupItem className="justify-content-between">
                    <Link to={`/edit/${id}`}>
                        <h3>{description}</h3>
                    </Link>
                    <i>Date: </i>{moment(createdAt).format('MMMM Do, YYYY')}
                    <Badge pill>
                        <Label>Amount: </Label>
                        {numeral(amount / 100).format('$0,0.00')}
                    </Badge>
                </ListGroupItem>
            </ListGroup>
        </div>
    );
};

export default ExpenseListItem;
