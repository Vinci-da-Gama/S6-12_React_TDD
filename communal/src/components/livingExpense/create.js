import React, { Component } from 'react';

class CreateExpense extends Component {
    render () {
        return (
            <div>
                create exp....
            </div>
        );
    }
};

export default CreateExpense;