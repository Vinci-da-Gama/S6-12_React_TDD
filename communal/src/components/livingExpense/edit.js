import React, { Component } from 'react';

class EditExpense extends Component {
    render () {
        return (
            <div>
                edit exp....
            </div>
        );
    }
};

export default EditExpense;