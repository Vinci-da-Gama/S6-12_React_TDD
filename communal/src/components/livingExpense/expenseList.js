import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button, Alert } from 'reactstrap';

import selectExpenses from '../../helpers/data-handler/expense-filterNsort';
import ExpenseListItem from './Expense_ListItem';

export class ExpenseListCompo extends Component {
    render() {
        if (this.props.expenses.length === 0 || !this.props.expenses) {
            return (
                <Alert color="danger">
                    No expense item list.
                </Alert>
            );
        }
        return (
            <div>
                {
                    this.props.expenses.map((expense, idx) => {
                        return <ExpenseListItem key={ expense.id+idx } {...expense} />
                    })
                }
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.setValViaFilters)
});

/* const mapDispatchToProps = { 
} */

export default connect(mapStateToProps, null)(ExpenseListCompo);
