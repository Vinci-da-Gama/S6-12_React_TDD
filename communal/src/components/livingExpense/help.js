import React, { Component } from 'react';

class HelpExpense extends Component {
    render () {
        return (
            <div>
                help exp....
            </div>
        );
    }
};

export default HelpExpense;