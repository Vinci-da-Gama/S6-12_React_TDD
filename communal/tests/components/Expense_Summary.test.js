import React from 'react';
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../src/components/livingExpense/Expense_Summary';
import fixtureSummary from '../fixtures/summary.fixture';

let w1;

beforeEach(() => {
    w1 = shallow(<ExpensesSummary expenseCount={fixtureSummary[0]} expensesTotal={fixtureSummary[0]} />);
});

test('should render expense summary with single expense.', () => {
    expect(w1).toMatchSnapshot();
});

test('should render expense summary with multiple expenses.', () => {
    const w2 = shallow(<ExpensesSummary expenseCount={fixtureSummary[1]} expensesTotal={fixtureSummary[1]} />);
    expect(w2).toMatchSnapshot();
});