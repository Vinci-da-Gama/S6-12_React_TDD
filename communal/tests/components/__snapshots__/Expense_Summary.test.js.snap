// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary with multiple expenses. 1`] = `
<div
  className="bg-success"
>
  Viewing 
  <Component />
   
  expenses
   totalling 
  $0.00
</div>
`;

exports[`should render expense summary with single expense. 1`] = `
<div
  className="bg-success"
>
  Viewing 
  <Component />
   
  expenses
   totalling 
  $0.00
</div>
`;
